File: android/src/main/java/me/heyteacher/flutter_antplus/heartrate/HeartrateHostApiImpl.java
package me.heyteacher.flutter_antplus.heartrate;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.dsi.ant.plugins.antplus.pcc.AntPlusHeartRatePcc;
import com.dsi.ant.plugins.antplus.pcc.defines.DeviceState;
import com.dsi.ant.plugins.antplus.pcc.defines.RequestAccessResult;
import com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc;
import com.dsi.ant.plugins.antplus.pccbase.AsyncScanController;
import com.dsi.ant.plugins.antplus.pccbase.PccReleaseHandle;

import io.flutter.Log;
import io.flutter.plugin.common.BinaryMessenger;

public class HeartrateHostApiImpl implements HeartratePigeon.HeartrateHostApi {

    private static final String TAG = HeartrateHostApiImpl.class.getSimpleName();
    private final Context context;
    private final HeartratePigeon.HeartrateFlutterApi flutterApi;
    private final HeartratePigeon.VoidResult flutterApiResult = new HeartratePigeon.VoidResult() {
        @Override
        public void success() {
        }

        @Override
        public void error(@NonNull Throwable error) {
            Log.e(TAG, "<onSearchResult>: error", error);
        }
    };
    private PccReleaseHandle<AntPlusHeartRatePcc> releaseHandle = null;
    private AntPlusHeartRatePcc hrPcc;
    private final AntPluginPcc.IDeviceStateChangeReceiver stateChangeReceiver = new AntPluginPcc.IDeviceStateChangeReceiver() {

        @Override
        public void onDeviceStateChange(DeviceState deviceState) {
            Log.d(TAG, "<onDeviceStateChange>: " + hrPcc.getAntDeviceNumber() + " deviceState " + deviceState);
            new Handler(Looper.getMainLooper()).post(() -> flutterApi.onDeviceStateChange(switch (deviceState) {
                case DEAD -> HeartratePigeon.AntplusDeviceState.DEAD;
                case CLOSED -> HeartratePigeon.AntplusDeviceState.CLOSED;
                case PROCESSING_REQUEST -> HeartratePigeon.AntplusDeviceState.PROCESSING_REQUEST;
                case SEARCHING -> HeartratePigeon.AntplusDeviceState.SEARCHING;
                case TRACKING -> HeartratePigeon.AntplusDeviceState.TRACKING;
                case UNRECOGNIZED -> HeartratePigeon.AntplusDeviceState.UNRECOGNIZED;
            }, flutterApiResult));
        }
    };
    private final AntPluginPcc.IPluginAccessResultReceiver<AntPlusHeartRatePcc> accessResultReceiver = new AntPluginPcc.IPluginAccessResultReceiver<>() {
        //Handle the result, connecting to events on success or reporting failure to user.
        @Override
        public void onResultReceived(AntPlusHeartRatePcc result, RequestAccessResult resultCode, DeviceState initialDeviceState) {
            Log.d(TAG, "<onResultReceived>: resultCode " + resultCode);
            switch (resultCode) {
                case SUCCESS:
                    hrPcc = result;
                    Log.d(TAG, "(onResultReceived): resultCode " + resultCode + ". deviceNumber " + result.getAntDeviceNumber() + " initialDeviceState " + initialDeviceState);
                    new Handler(Looper.getMainLooper()).post(() -> flutterApi.onDeviceStateChange(switch (initialDeviceState) {
                        case DEAD -> HeartratePigeon.AntplusDeviceState.DEAD;
                        case CLOSED -> HeartratePigeon.AntplusDeviceState.CLOSED;
                        case PROCESSING_REQUEST -> HeartratePigeon.AntplusDeviceState.PROCESSING_REQUEST;
                        case SEARCHING -> HeartratePigeon.AntplusDeviceState.SEARCHING;
                        case TRACKING -> HeartratePigeon.AntplusDeviceState.TRACKING;
                        case UNRECOGNIZED -> HeartratePigeon.AntplusDeviceState.UNRECOGNIZED;
                    }, flutterApiResult));
                    hrPcc.subscribeHeartRateDataEvent((estTimestamp, eventFlags, computedHeartRate, heartBeatCount, heartBeatEventTime, dataState) -> new Handler(Looper.getMainLooper()).post(() -> flutterApi.onHeartRateData((long) computedHeartRate, flutterApiResult)));
                    break;
                case DEPENDENCY_NOT_INSTALLED:
                    Log.e(TAG, "(onResultReceived): resultCode " + resultCode + ". The required service\"" + AntPlusHeartRatePcc.getMissingDependencyName() + "\"was not found. You need to install the ANT+ Plugins service or you may need to update your existing version if you already have it. Do you want to launch the Play Store to get it?");
                    break;
                case UNRECOGNIZED:
                case CHANNEL_NOT_AVAILABLE:
                case ADAPTER_NOT_DETECTED:
                case BAD_PARAMS:
                case OTHER_FAILURE:
                case USER_CANCELLED:
                default:
            }
        }
    };
    private AsyncScanController<AntPlusHeartRatePcc> hrScanCtrl;

    public HeartrateHostApiImpl(Context context, BinaryMessenger binaryMessenger) {
        this.context = context;
        this.flutterApi = new HeartratePigeon.HeartrateFlutterApi(binaryMessenger);
        HeartratePigeon.HeartrateHostApi.setUp(binaryMessenger, this);
    }

    @Override
    public void startScan() {
        Log.d(TAG, "<startScan>:");
        hrScanCtrl = AntPlusHeartRatePcc.requestAsyncScanController(context, 0, new AsyncScanController.IAsyncScanResultReceiver() {
            @Override
            public void onSearchStopped(RequestAccessResult reasonStopped) {
                Log.d(TAG, "<onSearchStopped>:");
                //The triggers calling this function use the same codes and require the same actions as those received by the standard access result receiver
                accessResultReceiver.onResultReceived(null, reasonStopped, DeviceState.DEAD);
            }

            @Override
            public void onSearchResult(final AsyncScanController.AsyncScanResultDeviceInfo deviceFound) {
                Log.d(TAG, "<onSearchResult>: number " + deviceFound.getAntDeviceNumber() + " name " + deviceFound.getDeviceDisplayName() + " already connected " + deviceFound.isAlreadyConnected());
                HeartratePigeon.AntplusDevice.Builder builder = new HeartratePigeon.AntplusDevice.Builder().setName(deviceFound.getDeviceDisplayName()).setNumber((long) deviceFound.getAntDeviceNumber());
                new Handler(Looper.getMainLooper()).post(() -> flutterApi.onScanResult(builder.build(), flutterApiResult));
            }
        });
    }

    @Override
    public void stopScan() {
        if (hrScanCtrl != null) {
            hrScanCtrl.closeScanController();
        }
        hrScanCtrl = null;
    }

    @Override
    public void connect(Long deviceNumber) {
        Log.d(TAG, "<connect>: deviceNumber " + deviceNumber);
        stopScan();
        releaseHandle = AntPlusHeartRatePcc.requestAccess(context, deviceNumber.intValue(), -1, accessResultReceiver, stateChangeReceiver);
    }

    @Override
    public void disconnect() {
        Log.d(TAG, "<disconnect>:");
        if (releaseHandle != null) {
            releaseHandle.close();
        }
        if (hrPcc != null) {
            Log.d(TAG, "(disconnect): deviceNumber " + hrPcc.getAntDeviceNumber());
            hrPcc.subscribeHeartRateDataEvent(null);
            hrPcc.releaseAccess();
        }
        releaseHandle = null;
        hrPcc = null;
    }
}

File: android/src/main/kotlin/me/heyteacher/flutter_antplus/heartrate/HeartrateHostApiImpl.kt
package me.heyteacher.flutter_antplus.heartrate

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.dsi.ant.plugins.antplus.pcc.AntPlusHeartRatePcc
import com.dsi.ant.plugins.antplus.pcc.AntPlusHeartRatePcc.DataState
import com.dsi.ant.plugins.antplus.pcc.defines.DeviceState
import com.dsi.ant.plugins.antplus.pcc.defines.EventFlag
import com.dsi.ant.plugins.antplus.pcc.defines.RequestAccessResult
import com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc.IDeviceStateChangeReceiver
import com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc.IPluginAccessResultReceiver
import com.dsi.ant.plugins.antplus.pccbase.AsyncScanController
import com.dsi.ant.plugins.antplus.pccbase.AsyncScanController.AsyncScanResultDeviceInfo
import com.dsi.ant.plugins.antplus.pccbase.AsyncScanController.IAsyncScanResultReceiver
import com.dsi.ant.plugins.antplus.pccbase.PccReleaseHandle
import io.flutter.plugin.common.BinaryMessenger
import me.heyteacher.flutter_antplus.device.OnDeviceStateChangeListener
import me.heyteacher.flutter_antplus.device.OnRequestAccessResultListener
import me.heyteacher.flutter_antplus.device.OnScanResultListener
import me.heyteacher.flutter_antplus.device.pigeons.AntplusDevice
import me.heyteacher.flutter_antplus.device.pigeons.AntplusDeviceType
import me.heyteacher.flutter_antplus.device.pigeons.OnDeviceStateChangeStreamHandler
import me.heyteacher.flutter_antplus.device.pigeons.OnRequestAccessResultStreamHandler
import me.heyteacher.flutter_antplus.device.pigeons.OnScanResultStreamHandler
import me.heyteacher.flutter_antplus.heartrate.pigeons.HeartrateHostApi
import me.heyteacher.flutter_antplus.heartrate.pigeons.OnHeartRateDataStreamHandler
import me.heyteacher.flutter_antplus.logging.OnLogDataListener
import me.heyteacher.flutter_antplus.logging.pigeons.AntplusLogData
import me.heyteacher.flutter_antplus.logging.pigeons.AntplusLogEvent
import java.math.BigDecimal
import java.util.EnumSet
import me.heyteacher.flutter_antplus.device.pigeons.AntplusDeviceState as PigeonDeviceState
import me.heyteacher.flutter_antplus.device.pigeons.AntplusRequestAccessResult as PigeonRequestAccessResult

class HeartrateHostApiImpl(
    private val context: Context,
    binaryMessenger: BinaryMessenger,
    private var onLogDataListener: OnLogDataListener
) : HeartrateHostApi {

    companion object {
        private val TAG: String = HeartrateHostApiImpl::class.java.simpleName
    }

    private var onScanResultListener: OnScanResultListener = OnScanResultListener()
    private var onDeviceStateChangeListener: OnDeviceStateChangeListener =
        OnDeviceStateChangeListener()
    private var onRequestAccessResultListener: OnRequestAccessResultListener =
        OnRequestAccessResultListener()
    private var onHeartRateDataListener: OnHeartRateDataListener = OnHeartRateDataListener()

    private var releaseHandle: PccReleaseHandle<AntPlusHeartRatePcc>? = null
    private var hrPcc: AntPlusHeartRatePcc? = null

    private val stateChangeReceiver = IDeviceStateChangeReceiver { deviceState ->
        Handler(Looper.getMainLooper()).post {
            onLogDataListener.add(
                AntplusLogData(
                    AntplusLogEvent.VERBOSE,
                    TAG,
                    "<onDeviceStateChange>: " + hrPcc?.antDeviceNumber + " deviceState " + deviceState
                )
            )
            onDeviceStateChangeListener.add(PigeonDeviceState.valueOf(deviceState.toString()))
        }
    }
    private val accessResultReceiver: IPluginAccessResultReceiver<AntPlusHeartRatePcc?> =
        IPluginAccessResultReceiver { result, resultCode, initialDeviceState ->
            when (resultCode) {
                RequestAccessResult.SUCCESS -> {
                    hrPcc = result
                    hrPcc!!.subscribeHeartRateDataEvent { _: Long, _: EnumSet<EventFlag?>?, computedHeartRate: Int, _: Long, _: BigDecimal?, _: DataState? ->
                        Handler(Looper.getMainLooper()).post {
                            onHeartRateDataListener.add(computedHeartRate.toLong())
                        }
                    }
                    Handler(Looper.getMainLooper()).post {
                        onRequestAccessResultListener.add(
                            PigeonRequestAccessResult.valueOf(resultCode.name)
                        )
                        onDeviceStateChangeListener.add(
                            PigeonDeviceState.valueOf(
                                initialDeviceState.toString()
                            )
                        )
                        onLogDataListener.add(
                            AntplusLogData(
                                AntplusLogEvent.INFO,
                                TAG,
                                "(onResultReceived): resultCode $resultCode. deviceNumber " + (result?.antDeviceNumber
                                    ?: "") + " initialDeviceState " + initialDeviceState
                            )
                        )
                    }
                }

                RequestAccessResult.DEPENDENCY_NOT_INSTALLED -> Handler(Looper.getMainLooper()).post {
                    onRequestAccessResultListener.add(
                        PigeonRequestAccessResult.valueOf(resultCode.name)
                    )
                    onLogDataListener.add(
                        AntplusLogData(
                            AntplusLogEvent.ERROR,
                            TAG,
                            "(onResultReceived): resultCode " + resultCode + ". The required service\"" + AntPlusHeartRatePcc.getMissingDependencyName() + "\"was not found. You need to install the ANT+ Plugins service or you may need to update your existing version if you already have it. Do you want to launch the Play Store to get it?"
                        )
                    )
                }

                else -> {
                    Handler(Looper.getMainLooper()).post {
                        onRequestAccessResultListener.add(
                            PigeonRequestAccessResult.valueOf(resultCode.name)
                        )
                        onLogDataListener.add(
                            AntplusLogData(
                                AntplusLogEvent.WARNING, TAG, "<onResultReceived>: resultCode $resultCode"
                            )
                        )
                    }
                }
            }
        }
    private var hrScanCtrl: AsyncScanController<AntPlusHeartRatePcc>? = null

    /**
     * Initializes
     */
    init {
        HeartrateHostApi.setUp(binaryMessenger, this)
        OnHeartRateDataStreamHandler.register(binaryMessenger, onHeartRateDataListener)
        OnDeviceStateChangeStreamHandler.register(
            binaryMessenger, onDeviceStateChangeListener, AntplusDeviceType.HEARTRATE.name
        )
        OnScanResultStreamHandler.register(
            binaryMessenger, onScanResultListener, AntplusDeviceType.HEARTRATE.name
        )
        OnRequestAccessResultStreamHandler.register(
            binaryMessenger, onRequestAccessResultListener, AntplusDeviceType.HEARTRATE.name
        )
    }

    /**
     * Starts the scan
     */
    override fun startScan() {
        Handler(Looper.getMainLooper()).post {
            onLogDataListener.add(
                AntplusLogData(
                    AntplusLogEvent.VERBOSE, TAG, "<startScan>:"
                )
            )
        }
        hrScanCtrl = AntPlusHeartRatePcc.requestAsyncScanController(
            context,
            0,
            object : IAsyncScanResultReceiver {
                override fun onSearchStopped(reasonStopped: RequestAccessResult) {
                    Handler(Looper.getMainLooper()).post {
                        onLogDataListener.add(
                            AntplusLogData(
                                AntplusLogEvent.VERBOSE, TAG, "<onSearchStopped>:"
                            )
                        )
                    }
                    //The triggers calling this function use the same codes and require the same actions as those received by the standard access result receiver
                    accessResultReceiver.onResultReceived(null, reasonStopped, DeviceState.DEAD)
                }

                override fun onSearchResult(deviceFound: AsyncScanResultDeviceInfo) {
                    Handler(Looper.getMainLooper()).post {
                        onLogDataListener.add(
                            AntplusLogData(
                                AntplusLogEvent.VERBOSE,
                                TAG,
                                "<onSearchResult>: number " + deviceFound.antDeviceNumber + " name " + deviceFound.deviceDisplayName + " already connected " + deviceFound.isAlreadyConnected
                            )
                        )
                        onScanResultListener.add(
                            AntplusDevice(
                                deviceFound.antDeviceNumber.toLong(),
                                deviceFound.deviceDisplayName,
                                AntplusDeviceType.HEARTRATE
                            )
                        )
                    }
                }
            })
    }

    /**
     * Stops the scan
     */
    override fun stopScan() {
        Handler(Looper.getMainLooper()).post {
            onLogDataListener.add(
                AntplusLogData(
                    AntplusLogEvent.VERBOSE, TAG, "<stopScan>:"
                )
            )
        }
        if (hrScanCtrl != null) {
            hrScanCtrl!!.closeScanController()
        }
        hrScanCtrl = null
    }

    /**
     * Connects to device with device number specified
     * @param deviceNumber the device number
     */
    override fun connect(deviceNumber: Long) {
        Handler(Looper.getMainLooper()).post {
            onLogDataListener.add(
                AntplusLogData(
                    AntplusLogEvent.VERBOSE, TAG, "<connect>: deviceNumber $deviceNumber"
                )
            )
        }
        stopScan()
        releaseHandle = AntPlusHeartRatePcc.requestAccess(
            context, deviceNumber.toInt(), -1, accessResultReceiver, stateChangeReceiver
        )
    }

    /**
     * Disconnects the device
     */
    override fun disconnect() {
        Handler(Looper.getMainLooper()).post {
            onLogDataListener.add(
                AntplusLogData(
                    AntplusLogEvent.VERBOSE, TAG, "<disconnect>:"
                )
            )
        }
        if (releaseHandle != null) {
            releaseHandle!!.close()
        }
        if (hrPcc != null) {
            Handler(Looper.getMainLooper()).post {
                onLogDataListener.add(
                    AntplusLogData(
                        AntplusLogEvent.VERBOSE, TAG, "(disconnect): deviceNumber " + hrPcc?.antDeviceNumber
                    )
                )
            }
            hrPcc?.subscribeHeartRateDataEvent(null)
            hrPcc?.releaseAccess()
        }
        releaseHandle = null
        hrPcc = null
    }

    /**
     * Closes listeners
     */
    fun close() {
        // closes device listeners
        onScanResultListener.onEventsDone()
        onRequestAccessResultListener.onEventsDone()
        onDeviceStateChangeListener.onEventsDone()
        // closes heartrate listener
        onHeartRateDataListener.onEventsDone()
    }
}